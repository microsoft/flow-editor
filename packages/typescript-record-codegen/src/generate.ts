// eslint-disable-next-line no-restricted-imports
import { template, TemplateExecutor } from "lodash";
import * as path from "path";
import {
  ClassDeclaration,
  InterfaceDeclaration,
  Project,
  PropertySignature,
  Scope,
  SourceFile,
  WriterFunction,
} from "ts-morph";
import { TS_RECORD } from "./constants";

function getPropertyType(property: PropertySignature): string | WriterFunction {
  const structure = property.getStructure();
  const propertyType = property.getType();
  if (!property.hasQuestionToken()) {
    return structure.type ?? "unknown";
  } else if (propertyType.isUnion()) {
    const unionTypes = propertyType.getUnionTypes();
    if (unionTypes.every((it) => !it.isUndefined())) {
      return `${structure.type} | undefined`;
    } else {
      return unionTypes.map((type) => type.getText()).join(" | ");
    }
  } else if (!propertyType.isUndefined()) {
    return `(${structure.type}) | undefined`;
  }
  return structure.type ?? "unknown";
}

function processClass(klass: ClassDeclaration, src: InterfaceDeclaration) {
  src.getProperties().forEach((property) => {
    const type = getPropertyType(property);
    klass.addProperty({
      name: property.getName(),
      isReadonly: true,
      hasQuestionToken: false,
      scope: Scope.Public,
      type,
    });
  });
}

export interface IGenerateOptions {
  sources: string[];
  outDir: string;
  tsConfig: string;
  classNameTemplate: string;
}

function transformName(
  src: InterfaceDeclaration,
  nameTemplate: TemplateExecutor
): string {
  const name = src.getName().match(/^I(.+)$/)?.[1] ?? "Unknown";
  return nameTemplate({
    name,
  });
}

export function generate(options: IGenerateOptions): SourceFile[] {
  const nameTemplate = template(options.classNameTemplate);
  const project = new Project({
    tsConfigFilePath: options.tsConfig,
  });
  const files: SourceFile[] = [];
  options.sources.forEach((sourcePath) => {
    const source = project.addSourceFileAtPath(sourcePath);
    if (!source) {
      return;
    }
    let target: SourceFile | undefined;
    source.getInterfaces().forEach((singleInterface) => {
      const jsDocs = singleInterface.getJsDocs();
      let shouldProcess = false;
      jsDocs.forEach((jsDoc) => {
        jsDoc.getTags().forEach((tag) => {
          switch (tag.getTagName()) {
            case TS_RECORD:
              shouldProcess = true;
              break;
            default:
              break;
          }
        });
      });

      if (!shouldProcess) {
        return;
      }
      if (!target) {
        const sourceFilename = source.getBaseNameWithoutExtension();
        const targetFilename = `${sourceFilename.split(".")[0]}.record.ts`;
        target = project.createSourceFile(
          path.resolve(options.outDir, targetFilename),
          "",
          {
            overwrite: true,
          }
        );
      }
      if (!target) {
        return;
      }
      const klass = target.addClass({
        name: transformName(singleInterface, nameTemplate),
      });
      processClass(klass, singleInterface);
    });
    if (!target) {
      return;
    }
    files.push(target);
  });
  return files;
}
