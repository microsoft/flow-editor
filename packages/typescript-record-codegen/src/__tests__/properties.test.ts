import * as path from "path";
import { generate } from "../generate";

it("should generate properties correctly", async () => {
  const files = generate({
    tsConfig: path.resolve(__dirname, "../../tsconfig.spec.json"),
    sources: [path.resolve(__dirname, "./input.interface.ts")],
    outDir: path.resolve(__dirname, "__generated__"),
    classNameTemplate: "<%= name %>Record",
  });
  files.forEach((file) => {
    const text = file.getText();
    expect(text).toMatchSnapshot(file.getBaseName());
  });
});
