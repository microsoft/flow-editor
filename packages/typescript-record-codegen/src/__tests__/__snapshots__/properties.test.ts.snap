// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate properties correctly: input.record.ts 1`] = `
"class InputRecord {
    public readonly simple: boolean;
    public readonly simpleOptional: undefined | symbol;
    public readonly nullable: undefined | null | string;
    public readonly hasUndefinedUnion: number | undefined;
}
"
`;
